{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## find gene overlap between monkey data and mouse references"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## mouse data vs mouse references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "\n",
    "data_type = 'float32'\n",
    "\n",
    "\n",
    "\n",
    "import matplotlib as mpl\n",
    "from matplotlib import rcParams\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# silence scanpy that prints a lot of warnings\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "rcParams['pdf.fonttype'] = 42 # enables correct plotting of text for PDFs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_folder = '/home/ssobti/projects/farnaz_spatial/output_data/monkey/cell2location/multi_ref_annot'\n",
    "\n",
    "# create paths and names to results folders for reference regression and cell2location models\n",
    "ref_run_name = f'{results_folder}/reference_signatures'\n",
    "#run_name = f'{results_folder}/cell2location_map'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## borrow the combined reference from the mouse data\n",
    "combined_refs = anndata.read_h5ad('/home/ssobti/projects/farnaz_spatial/data/mouse_data/combined_and_independent_refs/combined_refs_raw_data_seurat.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "## load in mouse data\n",
    "# Set paths to data and results used through the document:\n",
    "sp_data_folder = '/home/ssobti/projects/farnaz_spatial/data/mouse_data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_and_qc(sample_name, path=sp_data_folder):\n",
    "    r\"\"\" This function reads the data for one 10X spatial experiment into the anndata object.\n",
    "    It also calculates QC metrics. Modify this function if required by your workflow.\n",
    "\n",
    "    :param sample_name: Name of the sample\n",
    "    :param path: path to data\n",
    "    \"\"\"\n",
    "\n",
    "    adata = sc.read_visium(path + str(sample_name) + '/outs',\n",
    "                           count_file='filtered_feature_bc_matrix.h5', load_images=True)\n",
    "    adata.obs['sample'] = sample_name\n",
    "    adata.var['SYMBOL'] = adata.var_names\n",
    "    adata.var.rename(columns={'gene_ids': 'ENSEMBL'}, inplace=True)\n",
    "    adata.var_names = adata.var['ENSEMBL']\n",
    "    adata.var.drop(columns='ENSEMBL', inplace=True)\n",
    "\n",
    "    # Calculate QC metrics\n",
    "    from scipy.sparse import csr_matrix\n",
    "    adata.X = adata.X.toarray()\n",
    "    sc.pp.calculate_qc_metrics(adata, inplace=True)\n",
    "    adata.X = csr_matrix(adata.X)\n",
    "    adata.var['mt'] = [gene.startswith('mt-') for gene in adata.var['SYMBOL']]\n",
    "    adata.obs['mt_frac'] = adata[:, adata.var['mt'].tolist()].X.sum(1).A.squeeze()/adata.obs['total_counts']\n",
    "\n",
    "    # add sample name to obs names\n",
    "    adata.obs[\"sample\"] = [str(i) for i in adata.obs['sample']]\n",
    "    adata.obs_names = adata.obs[\"sample\"] \\\n",
    "                          + '_' + adata.obs_names\n",
    "    adata.obs.index.name = 'spot_id'\n",
    "\n",
    "    return adata\n",
    "\n",
    "def select_slide(adata, s, s_col='sample'):\n",
    "    r\"\"\" This function selects the data for one slide from the spatial anndata object.\n",
    "\n",
    "    :param adata: Anndata object with multiple spatial experiments\n",
    "    :param s: name of selected experiment\n",
    "    :param s_col: column in adata.obs listing experiment name for each location\n",
    "    \"\"\"\n",
    "\n",
    "    slide = adata[adata.obs[s_col].isin([s]), :]\n",
    "    s_keys = list(slide.uns['spatial'].keys())\n",
    "    s_spatial = np.array(s_keys)[[s in k for k in s_keys]][0]\n",
    "\n",
    "    slide.uns['spatial'] = {s_spatial: slide.uns['spatial'][s_spatial]}\n",
    "\n",
    "    return slide\n",
    "\n",
    "#######################\n",
    "# Read the list of spatial experiments\n",
    "sample_data = pd.read_csv(sp_data_folder + 'visium.csv')\n",
    "\n",
    "# Read the data into anndata objects\n",
    "slides = []\n",
    "for i in sample_data['sample_name']:\n",
    "    slides.append(read_and_qc(i, path=sp_data_folder))\n",
    "\n",
    "# Combine anndata objects together\n",
    "adata_vis = slides[0].concatenate(\n",
    "    slides[1:],\n",
    "    batch_key=\"sample\",\n",
    "    uns_merge=\"unique\",\n",
    "    batch_categories=sample_data['sample_name'],\n",
    "    index_unique=None\n",
    ")\n",
    "#######################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14097"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(combined_refs.var['features']).intersection(set(adata_vis.var['SYMBOL'])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## monkey data vs mouse references"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set paths to data and results used through the document:\n",
    "sp_data_folder = '/home/ssobti/projects/farnaz_spatial/data/monkey_data/unpacked_data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_and_qc(sample_name, path=sp_data_folder):\n",
    "    r\"\"\" This function reads the data for one 10X spatial experiment into the anndata object.\n",
    "    It also calculates QC metrics. Modify this function if required by your workflow.\n",
    "\n",
    "    :param sample_name: Name of the sample\n",
    "    :param path: path to data\n",
    "    \"\"\"\n",
    "\n",
    "    adata = sc.read_visium(path + str(sample_name) + '/outs',\n",
    "                           count_file='filtered_feature_bc_matrix.h5', load_images=True)\n",
    "    adata.obs['sample'] = sample_name\n",
    "    adata.var['SYMBOL'] = adata.var_names\n",
    "    adata.var.rename(columns={'gene_ids': 'ENSEMBL'}, inplace=True)\n",
    "    adata.var_names = adata.var['ENSEMBL']\n",
    "    adata.var.drop(columns='ENSEMBL', inplace=True)\n",
    "\n",
    "    # Calculate QC metrics\n",
    "    from scipy.sparse import csr_matrix\n",
    "    adata.X = adata.X.toarray()\n",
    "    sc.pp.calculate_qc_metrics(adata, inplace=True)\n",
    "    adata.X = csr_matrix(adata.X)\n",
    "    adata.var['mt'] = [gene.startswith('mt-') for gene in adata.var['SYMBOL']]\n",
    "    adata.obs['mt_frac'] = adata[:, adata.var['mt'].tolist()].X.sum(1).A.squeeze()/adata.obs['total_counts']\n",
    "\n",
    "    # add sample name to obs names\n",
    "    adata.obs[\"sample\"] = [str(i) for i in adata.obs['sample']]\n",
    "    adata.obs_names = adata.obs[\"sample\"] \\\n",
    "                          + '_' + adata.obs_names\n",
    "    adata.obs.index.name = 'spot_id'\n",
    "\n",
    "    return adata\n",
    "\n",
    "def select_slide(adata, s, s_col='sample'):\n",
    "    r\"\"\" This function selects the data for one slide from the spatial anndata object.\n",
    "\n",
    "    :param adata: Anndata object with multiple spatial experiments\n",
    "    :param s: name of selected experiment\n",
    "    :param s_col: column in adata.obs listing experiment name for each location\n",
    "    \"\"\"\n",
    "\n",
    "    slide = adata[adata.obs[s_col].isin([s]), :]\n",
    "    s_keys = list(slide.uns['spatial'].keys())\n",
    "    s_spatial = np.array(s_keys)[[s in k for k in s_keys]][0]\n",
    "\n",
    "    slide.uns['spatial'] = {s_spatial: slide.uns['spatial'][s_spatial]}\n",
    "\n",
    "    return slide\n",
    "\n",
    "#######################\n",
    "# Read the list of spatial experiments\n",
    "sample_data = pd.read_csv(sp_data_folder + 'visium.csv')\n",
    "\n",
    "# Read the data into anndata objects\n",
    "slides = []\n",
    "for i in sample_data['sample_name']:\n",
    "    slides.append(read_and_qc(i, path=sp_data_folder))\n",
    "\n",
    "#### change the slide identity key in .uns so they aren't the repeats between slides ####\n",
    "## new names:\n",
    "## 096-count-A1: 1149_G\n",
    "## 096-count-D1: 1572_G\n",
    "## 327-count-A1: 1080_T\n",
    "## 327-count-D1: 1080_B\n",
    "\n",
    "slides[0].uns['spatial']['1149G_A1_096'] = slides[0].uns['spatial'].pop('count-A1')\n",
    "slides[1].uns['spatial']['1572G_D1_096'] = slides[1].uns['spatial'].pop('count-D1')\n",
    "slides[2].uns['spatial']['1080T_A1_327'] = slides[2].uns['spatial'].pop('count-A1')\n",
    "slides[3].uns['spatial']['1080B_D1_327'] = slides[3].uns['spatial'].pop('count-D1')\n",
    "\n",
    "# Combine anndata objects together\n",
    "adata_vis = slides[0].concatenate(\n",
    "    slides[1:],\n",
    "    batch_key=\"sample\",\n",
    "    uns_merge=\"unique\",\n",
    "    batch_categories=sample_data['sample_name'],\n",
    "    index_unique=None\n",
    ")\n",
    "#######################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "11"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(combined_refs.var['features']).intersection(set(adata_vis.var['SYMBOL'])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_types</th>\n",
       "      <th>genome</th>\n",
       "      <th>SYMBOL</th>\n",
       "      <th>mt</th>\n",
       "      <th>n_cells_by_counts-1080B_D1_327</th>\n",
       "      <th>mean_counts-1080B_D1_327</th>\n",
       "      <th>log1p_mean_counts-1080B_D1_327</th>\n",
       "      <th>pct_dropout_by_counts-1080B_D1_327</th>\n",
       "      <th>total_counts-1080B_D1_327</th>\n",
       "      <th>log1p_total_counts-1080B_D1_327</th>\n",
       "      <th>...</th>\n",
       "      <th>log1p_mean_counts-1149G_A1_096</th>\n",
       "      <th>pct_dropout_by_counts-1149G_A1_096</th>\n",
       "      <th>total_counts-1149G_A1_096</th>\n",
       "      <th>log1p_total_counts-1149G_A1_096</th>\n",
       "      <th>n_cells_by_counts-1572G_D1_096</th>\n",
       "      <th>mean_counts-1572G_D1_096</th>\n",
       "      <th>log1p_mean_counts-1572G_D1_096</th>\n",
       "      <th>pct_dropout_by_counts-1572G_D1_096</th>\n",
       "      <th>total_counts-1572G_D1_096</th>\n",
       "      <th>log1p_total_counts-1572G_D1_096</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSEMBL</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000187634</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>False</td>\n",
       "      <td>105</td>\n",
       "      <td>0.041590</td>\n",
       "      <td>0.040748</td>\n",
       "      <td>96.135444</td>\n",
       "      <td>113.0</td>\n",
       "      <td>4.736198</td>\n",
       "      <td>...</td>\n",
       "      <td>0.003692</td>\n",
       "      <td>99.768840</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.197225</td>\n",
       "      <td>119</td>\n",
       "      <td>0.038473</td>\n",
       "      <td>0.037751</td>\n",
       "      <td>96.423204</td>\n",
       "      <td>128.0</td>\n",
       "      <td>4.859812</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000188976</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NOC2L</td>\n",
       "      <td>False</td>\n",
       "      <td>82</td>\n",
       "      <td>0.031285</td>\n",
       "      <td>0.030805</td>\n",
       "      <td>96.981965</td>\n",
       "      <td>85.0</td>\n",
       "      <td>4.454347</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>59</td>\n",
       "      <td>0.018635</td>\n",
       "      <td>0.018464</td>\n",
       "      <td>98.226631</td>\n",
       "      <td>62.0</td>\n",
       "      <td>4.143135</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187961</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>KLHL17</td>\n",
       "      <td>False</td>\n",
       "      <td>104</td>\n",
       "      <td>0.040118</td>\n",
       "      <td>0.039334</td>\n",
       "      <td>96.172249</td>\n",
       "      <td>109.0</td>\n",
       "      <td>4.700480</td>\n",
       "      <td>...</td>\n",
       "      <td>0.006911</td>\n",
       "      <td>99.352751</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2.772589</td>\n",
       "      <td>177</td>\n",
       "      <td>0.055906</td>\n",
       "      <td>0.054399</td>\n",
       "      <td>94.679892</td>\n",
       "      <td>186.0</td>\n",
       "      <td>5.231109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187583</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>PLEKHN1</td>\n",
       "      <td>False</td>\n",
       "      <td>6</td>\n",
       "      <td>0.002208</td>\n",
       "      <td>0.002206</td>\n",
       "      <td>99.779168</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.945910</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000924</td>\n",
       "      <td>99.953768</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.098612</td>\n",
       "      <td>33</td>\n",
       "      <td>0.009919</td>\n",
       "      <td>0.009870</td>\n",
       "      <td>99.008115</td>\n",
       "      <td>33.0</td>\n",
       "      <td>3.526361</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000187642</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>PERM1</td>\n",
       "      <td>False</td>\n",
       "      <td>17</td>\n",
       "      <td>0.006257</td>\n",
       "      <td>0.006237</td>\n",
       "      <td>99.374310</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2.890372</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8</td>\n",
       "      <td>0.002405</td>\n",
       "      <td>0.002402</td>\n",
       "      <td>99.759543</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.197225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000212907</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>MT-ND4L</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001386</td>\n",
       "      <td>99.861304</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.386294</td>\n",
       "      <td>1</td>\n",
       "      <td>0.000301</td>\n",
       "      <td>0.000301</td>\n",
       "      <td>99.969943</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.693147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198886</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>MT-ND4</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000462</td>\n",
       "      <td>99.953768</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.693147</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198786</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>MT-ND5</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198695</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>MT-ND6</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000198727</th>\n",
       "      <td>Gene Expression</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>MT-CYB</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18085 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   feature_types  genome   SYMBOL     mt  \\\n",
       "ENSEMBL                                                    \n",
       "ENSG00000187634  Gene Expression  GRCh38   SAMD11  False   \n",
       "ENSG00000188976  Gene Expression  GRCh38    NOC2L  False   \n",
       "ENSG00000187961  Gene Expression  GRCh38   KLHL17  False   \n",
       "ENSG00000187583  Gene Expression  GRCh38  PLEKHN1  False   \n",
       "ENSG00000187642  Gene Expression  GRCh38    PERM1  False   \n",
       "...                          ...     ...      ...    ...   \n",
       "ENSG00000212907  Gene Expression  GRCh38  MT-ND4L  False   \n",
       "ENSG00000198886  Gene Expression  GRCh38   MT-ND4  False   \n",
       "ENSG00000198786  Gene Expression  GRCh38   MT-ND5  False   \n",
       "ENSG00000198695  Gene Expression  GRCh38   MT-ND6  False   \n",
       "ENSG00000198727  Gene Expression  GRCh38   MT-CYB  False   \n",
       "\n",
       "                 n_cells_by_counts-1080B_D1_327  mean_counts-1080B_D1_327  \\\n",
       "ENSEMBL                                                                     \n",
       "ENSG00000187634                             105                  0.041590   \n",
       "ENSG00000188976                              82                  0.031285   \n",
       "ENSG00000187961                             104                  0.040118   \n",
       "ENSG00000187583                               6                  0.002208   \n",
       "ENSG00000187642                              17                  0.006257   \n",
       "...                                         ...                       ...   \n",
       "ENSG00000212907                               0                  0.000000   \n",
       "ENSG00000198886                               0                  0.000000   \n",
       "ENSG00000198786                               0                  0.000000   \n",
       "ENSG00000198695                               0                  0.000000   \n",
       "ENSG00000198727                               0                  0.000000   \n",
       "\n",
       "                 log1p_mean_counts-1080B_D1_327  \\\n",
       "ENSEMBL                                           \n",
       "ENSG00000187634                        0.040748   \n",
       "ENSG00000188976                        0.030805   \n",
       "ENSG00000187961                        0.039334   \n",
       "ENSG00000187583                        0.002206   \n",
       "ENSG00000187642                        0.006237   \n",
       "...                                         ...   \n",
       "ENSG00000212907                        0.000000   \n",
       "ENSG00000198886                        0.000000   \n",
       "ENSG00000198786                        0.000000   \n",
       "ENSG00000198695                        0.000000   \n",
       "ENSG00000198727                        0.000000   \n",
       "\n",
       "                 pct_dropout_by_counts-1080B_D1_327  \\\n",
       "ENSEMBL                                               \n",
       "ENSG00000187634                           96.135444   \n",
       "ENSG00000188976                           96.981965   \n",
       "ENSG00000187961                           96.172249   \n",
       "ENSG00000187583                           99.779168   \n",
       "ENSG00000187642                           99.374310   \n",
       "...                                             ...   \n",
       "ENSG00000212907                          100.000000   \n",
       "ENSG00000198886                          100.000000   \n",
       "ENSG00000198786                          100.000000   \n",
       "ENSG00000198695                          100.000000   \n",
       "ENSG00000198727                          100.000000   \n",
       "\n",
       "                 total_counts-1080B_D1_327  log1p_total_counts-1080B_D1_327  \\\n",
       "ENSEMBL                                                                       \n",
       "ENSG00000187634                      113.0                         4.736198   \n",
       "ENSG00000188976                       85.0                         4.454347   \n",
       "ENSG00000187961                      109.0                         4.700480   \n",
       "ENSG00000187583                        6.0                         1.945910   \n",
       "ENSG00000187642                       17.0                         2.890372   \n",
       "...                                    ...                              ...   \n",
       "ENSG00000212907                        0.0                         0.000000   \n",
       "ENSG00000198886                        0.0                         0.000000   \n",
       "ENSG00000198786                        0.0                         0.000000   \n",
       "ENSG00000198695                        0.0                         0.000000   \n",
       "ENSG00000198727                        0.0                         0.000000   \n",
       "\n",
       "                 ...  log1p_mean_counts-1149G_A1_096  \\\n",
       "ENSEMBL          ...                                   \n",
       "ENSG00000187634  ...                        0.003692   \n",
       "ENSG00000188976  ...                        0.000000   \n",
       "ENSG00000187961  ...                        0.006911   \n",
       "ENSG00000187583  ...                        0.000924   \n",
       "ENSG00000187642  ...                        0.000000   \n",
       "...              ...                             ...   \n",
       "ENSG00000212907  ...                        0.001386   \n",
       "ENSG00000198886  ...                        0.000462   \n",
       "ENSG00000198786  ...                        0.000000   \n",
       "ENSG00000198695  ...                        0.000000   \n",
       "ENSG00000198727  ...                        0.000000   \n",
       "\n",
       "                 pct_dropout_by_counts-1149G_A1_096  \\\n",
       "ENSEMBL                                               \n",
       "ENSG00000187634                           99.768840   \n",
       "ENSG00000188976                          100.000000   \n",
       "ENSG00000187961                           99.352751   \n",
       "ENSG00000187583                           99.953768   \n",
       "ENSG00000187642                          100.000000   \n",
       "...                                             ...   \n",
       "ENSG00000212907                           99.861304   \n",
       "ENSG00000198886                           99.953768   \n",
       "ENSG00000198786                          100.000000   \n",
       "ENSG00000198695                          100.000000   \n",
       "ENSG00000198727                          100.000000   \n",
       "\n",
       "                 total_counts-1149G_A1_096  log1p_total_counts-1149G_A1_096  \\\n",
       "ENSEMBL                                                                       \n",
       "ENSG00000187634                        8.0                         2.197225   \n",
       "ENSG00000188976                        0.0                         0.000000   \n",
       "ENSG00000187961                       15.0                         2.772589   \n",
       "ENSG00000187583                        2.0                         1.098612   \n",
       "ENSG00000187642                        0.0                         0.000000   \n",
       "...                                    ...                              ...   \n",
       "ENSG00000212907                        3.0                         1.386294   \n",
       "ENSG00000198886                        1.0                         0.693147   \n",
       "ENSG00000198786                        0.0                         0.000000   \n",
       "ENSG00000198695                        0.0                         0.000000   \n",
       "ENSG00000198727                        0.0                         0.000000   \n",
       "\n",
       "                 n_cells_by_counts-1572G_D1_096  mean_counts-1572G_D1_096  \\\n",
       "ENSEMBL                                                                     \n",
       "ENSG00000187634                             119                  0.038473   \n",
       "ENSG00000188976                              59                  0.018635   \n",
       "ENSG00000187961                             177                  0.055906   \n",
       "ENSG00000187583                              33                  0.009919   \n",
       "ENSG00000187642                               8                  0.002405   \n",
       "...                                         ...                       ...   \n",
       "ENSG00000212907                               1                  0.000301   \n",
       "ENSG00000198886                               0                  0.000000   \n",
       "ENSG00000198786                               0                  0.000000   \n",
       "ENSG00000198695                               0                  0.000000   \n",
       "ENSG00000198727                               0                  0.000000   \n",
       "\n",
       "                 log1p_mean_counts-1572G_D1_096  \\\n",
       "ENSEMBL                                           \n",
       "ENSG00000187634                        0.037751   \n",
       "ENSG00000188976                        0.018464   \n",
       "ENSG00000187961                        0.054399   \n",
       "ENSG00000187583                        0.009870   \n",
       "ENSG00000187642                        0.002402   \n",
       "...                                         ...   \n",
       "ENSG00000212907                        0.000301   \n",
       "ENSG00000198886                        0.000000   \n",
       "ENSG00000198786                        0.000000   \n",
       "ENSG00000198695                        0.000000   \n",
       "ENSG00000198727                        0.000000   \n",
       "\n",
       "                 pct_dropout_by_counts-1572G_D1_096  \\\n",
       "ENSEMBL                                               \n",
       "ENSG00000187634                           96.423204   \n",
       "ENSG00000188976                           98.226631   \n",
       "ENSG00000187961                           94.679892   \n",
       "ENSG00000187583                           99.008115   \n",
       "ENSG00000187642                           99.759543   \n",
       "...                                             ...   \n",
       "ENSG00000212907                           99.969943   \n",
       "ENSG00000198886                          100.000000   \n",
       "ENSG00000198786                          100.000000   \n",
       "ENSG00000198695                          100.000000   \n",
       "ENSG00000198727                          100.000000   \n",
       "\n",
       "                 total_counts-1572G_D1_096  log1p_total_counts-1572G_D1_096  \n",
       "ENSEMBL                                                                      \n",
       "ENSG00000187634                      128.0                         4.859812  \n",
       "ENSG00000188976                       62.0                         4.143135  \n",
       "ENSG00000187961                      186.0                         5.231109  \n",
       "ENSG00000187583                       33.0                         3.526361  \n",
       "ENSG00000187642                        8.0                         2.197225  \n",
       "...                                    ...                              ...  \n",
       "ENSG00000212907                        1.0                         0.693147  \n",
       "ENSG00000198886                        0.0                         0.000000  \n",
       "ENSG00000198786                        0.0                         0.000000  \n",
       "ENSG00000198695                        0.0                         0.000000  \n",
       "ENSG00000198727                        0.0                         0.000000  \n",
       "\n",
       "[18085 rows x 28 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_vis.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gseapy as gp\n",
    "from gseapy import Biomart\n",
    "bm = Biomart()\n",
    "# note the dataset and attribute names are different\n",
    "m2h = bm.query(dataset='mmusculus_gene_ensembl',\n",
    "               attributes=['ensembl_gene_id','external_gene_name',\n",
    "                           'hsapiens_homolog_ensembl_gene',\n",
    "                           'hsapiens_homolog_associated_gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a dict symbol mappings\n",
    "m2h_dict = {}\n",
    "for i, row in m2h.loc[:,[\"external_gene_name\", \"hsapiens_homolog_associated_gene_name\"]].iterrows():\n",
    "    if row.isna().any(): continue\n",
    "    m2h_dict[row['external_gene_name']] = row[\"hsapiens_homolog_associated_gene_name\"]\n",
    "# read gmt file into dict\n",
    "#kegg = gp.read_gmt(path=\"tests/extdata/enrichr.KEGG_2016.gmt\")\n",
    "#print(kegg['MAPK signaling pathway Homo sapiens hsa04010'][:10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "m2h_genes_input = combined_refs.var['features']\n",
    "m2h_genes_output = []\n",
    "\n",
    "for gene in m2h_genes_input:\n",
    "    if gene in m2h_dict.keys():\n",
    "        m2h_genes_output.append(m2h_dict[gene])\n",
    "    else:\n",
    "        m2h_genes_output.append('NA')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16.40173410404624"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## percent of mouse genes that didn't map to human gene\n",
    "100*sum(np.array(m2h_genes_output) == 'NA')/len(m2h_genes_output)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "farnaz_spatial",
   "language": "python",
   "name": "farnaz_spatial"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
